### ISSUE TITLE
Process Freezing

### ISSUE THREAD

**Issue by kpheasey** [mperham/sidekiq#3337]
Ruby version:
Sidekiq / Pro / Enterprise version(s):
```
sidekiq (4.2.3)
sidekiq-pro (3.4.0)
```

Process continues to freeze sporadically.  After reading the frozen process section, I've ensured there is a timeout on all outgoing requests.  The max timeout is currently 600s.

I'm pretty sure it's related to FeederWorker, which is a class that scrapes external api's for users.  The other main job is for calculations which uses delayed extensions.  However, if I do not run feeds then the process never seems to freeze.

We run 2 processes per instance and autoscale the number of instances based on workload.  I've seen the process freeze even when running a single instance.

Here's a link to the gdb output - https://gist.github.com/kpheasey/2c987b4c23da90a40fb940138a6bf21e

Screenshot of the frozen process in web is below.  Note that there are 2 processes shown, one of which has frozen.

![screen shot 2017-01-31 at 9 41 21 am](https://cloud.githubusercontent.com/assets/3934075/22469305/8a0974f2-e799-11e6-9f3f-a6a3c7709492.png)

**Comment by mperham** [comment]
This:

```
#7  0x00007fa34eb8f0be in ?? () from /usr/lib/x86_64-linux-gnu/libruby-2.3.so.2.3
#8  0x00007fa34eb8fa5f in ?? () from /usr/lib/x86_64-linux-gnu/libruby-2.3.so.2.3
#9  0x00007fa34eb8fbd5 in ?? () from /usr/lib/x86_64-linux-gnu/libruby-2.3.so.2.3
#10 0x00007fa34eb8fd4c in ?? () from /usr/lib/x86_64-linux-gnu/libruby-2.3.so.2.3
```

Means you don't have Ruby debug symbols installed.  You'll need to install them for that GDB output to mean much.

I did see mysql mentioned, make sure you've got the latest of all native gems.

```
bundle exec ruby -e 'puts Gem.loaded_specs.values.select{ |i| !i.extensions.empty? }.map{ |i| i.name }'
```

**Comment by kpheasey** [comment]
Just wanted to respond with some findings, not re-opening the issue.

Re-compiling production ruby with necessary debugging symbols was not something we could do for debugging purposes.  We weren't able to replicate in development.

We did separate the the queues into different processes and found that calculations were the main cause.  Our calculation jobs cause a lot of deadlocks in the database writes because they touch associated record for cache invalidation.

Our solution was to process associated sets of calculations synchronously.  This lowered the overall level of concurrency, but by reducing database deadlocks are preventing process freezing, we increased our efficiency and reliability.
